The remove duplicate nodes option is now causing a SAXParseException.

When executing the command:
```
java -jar wadltest-2.5.0-jar-with-dependencies.jar -dpr simple-rbac.wadl
```
with the WADL file containing:
```
<application xmlns:rax="http://docs.rackspace.com/api"
         xmlns:xsd="http://www.w3.org/2001/XMLSchema"
         xmlns="http://wadl.dev.java.net/2009/02"
    >
  <resources base="http://localhost">
    <resource id="_0be44e11-68e6-44d4-acae-ec6a366282ba" path="/path/to/this">
      <method name="GET"    id="_0be44e11-68e6-44d4-acae-ec6a366282ba-GET"    rax:roles="role1 role2 role3 role4"/>
    </resource>
    <resource id="_0e126d22-d8d2-4ece-8667-b19875ae1e12" path="/path/to/this">
      <method name="PUT"    id="_0e126d22-d8d2-4ece-8667-b19875ae1e12-PUT"    rax:roles="role1 role2 role3"/>
    </resource>
    <resource id="_b3d2229d-cc90-4d53-9f05-c1c60117af07" path="/path/to/this">
      <method name="POST"   id="_b3d2229d-cc90-4d53-9f05-c1c60117af07-POST"   rax:roles="role1 role2"/>
    </resource>
    <resource id="_4724d9ab-d560-4668-bc18-3690cc7faa76" path="/path/to/this">
      <method name="DELETE" id="_4724d9ab-d560-4668-bc18-3690cc7faa76-DELETE" rax:roles="role1"/>
    </resource>
    <resource id="_bed37f49-c275-474b-8681-a3a48918ddb9" path="/path/to/that">
      <method name="GET"    id="_bed37f49-c275-474b-8681-a3a48918ddb9-GET"    />
      <method name="PUT"    id="_bed37f49-c275-474b-8681-a3a48918ddb9-PUT"    />
    </resource>
    <resource id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284" path="/path/to/that">
      <method name="GET"    id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284-GET"    rax:roles="role1"/>
      <method name="PUT"    id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284-PUT"    rax:roles="role1"/>
      <method name="POST"   id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284-POST"   rax:roles="role1"/>
      <method name="DELETE" id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284-DELETE" rax:roles="role1"/>
    </resource>
  </resources>
</application>
```
the following exception is received:
```
Loading file:/path/to/file/simple-rbac.wadl...

[ERROR] [main] org.xml.sax.SAXParseException; systemId: jar:file:/path/to/jar/wadltest-2.5.0-jar-with-dependencies.jar!/xsl/adjust-next-cont-error.xsl; cvc-id.1: There is no ID/IDREF binding for IDREF 'd43e25'.
Exception in thread "main" com.rackspace.com.papi.components.checker.wadl.WADLException: WADL Processing Error: org.xml.sax.SAXParseException; systemId: jar:file:/path/to/jar/wadltest-2.5.0-jar-with-dependencies.jar!/xsl/adjust-next-cont-error.xsl; cvc-id.1: There is no ID/IDREF binding for IDREF 'd43e25'.
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder.buildFromWADL(WADLCheckerBuilder.scala:707)
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder.build(WADLCheckerBuilder.scala:766)
        at com.rackspace.com.papi.components.checker.wadl.StepBuilder.build(StepBuilder.scala:59)
        at com.rackspace.com.papi.components.checker.Validator$.apply(Validator.scala:67)
        at com.rackspace.com.papi.components.checker.cli.WadlTest$.runServer(WadlTest.scala:210)
        at com.rackspace.com.papi.components.checker.cli.WadlTest$.main(WadlTest.scala:323)
        at com.rackspace.com.papi.components.checker.cli.WadlTest.main(WadlTest.scala)
Caused by: net.sf.saxon.trans.XPathException: org.xml.sax.SAXParseException; systemId: jar:file:/path/to/jar/wadltest-2.5.0-jar-with-dependencies.jar!/xsl/adjust-next-cont-error.xsl; cvc-id.1: There is no ID/IDREF binding for IDREF 'd43e25'.
        at net.sf.saxon.event.ContentHandlerProxy.handleSAXException(ContentHandlerProxy.java:555)
        at net.sf.saxon.event.ContentHandlerProxy.endElement(ContentHandlerProxy.java:416)
        at net.sf.saxon.event.NamespaceReducer.endElement(NamespaceReducer.java:228)
        at net.sf.saxon.event.ComplexContentOutputter.endElement(ComplexContentOutputter.java:576)
        at net.sf.saxon.event.ProxyReceiver.endElement(ProxyReceiver.java:182)
        at net.sf.saxon.tree.tiny.TinyElementImpl.copy(TinyElementImpl.java:414)
        at net.sf.saxon.tree.tiny.TinyDocumentImpl.copy(TinyDocumentImpl.java:287)
        at net.sf.saxon.event.Sender.sendDocumentInfo(Sender.java:243)
        at net.sf.saxon.event.Sender.send(Sender.java:126)
        at net.sf.saxon.jaxp.IdentityTransformer.transform(IdentityTransformer.java:374)
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder.com$rackspace$com$papi$components$checker$wadl$WADLCheckerBuilder$$validateChecker(WADLCheckerBuilder.scala:606)
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder$$anonfun$buildFromWADL$1$$anonfun$14.apply(WADLCheckerBuilder.scala:697)
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder$$anonfun$buildFromWADL$1$$anonfun$14.apply(WADLCheckerBuilder.scala:697)
        at com.rackspace.com.papi.components.checker.wadl.BuilderHelper$.applyBuildSteps(BuilderHelper.scala:129)
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder$$anonfun$buildFromWADL$1.apply$mcV$sp(WADLCheckerBuilder.scala:702)
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder$$anonfun$buildFromWADL$1.apply(WADLCheckerBuilder.scala:661)
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder$$anonfun$buildFromWADL$1.apply(WADLCheckerBuilder.scala:661)
        at com.rackspace.cloud.api.wadl.util.XSLErrorDispatcher$class.handleXSLException(xsl-error-dispatcher.scala:15)
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder.handleXSLException(WADLCheckerBuilder.scala:223)
        at com.rackspace.com.papi.components.checker.wadl.WADLCheckerBuilder.buildFromWADL(WADLCheckerBuilder.scala:661)
        ... 6 more
Caused by: org.xml.sax.SAXParseException; systemId: jar:file:/path/to/jar/wadltest-2.5.0-jar-with-dependencies.jar!/xsl/adjust-next-cont-error.xsl; cvc-id.1: There is no ID/IDREF binding for IDREF 'd43e25'.
        at org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:196)
        at org.apache.xerces.util.ErrorHandlerWrapper.error(ErrorHandlerWrapper.java:132)
        at org.apache.xerces.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:394)
        at org.apache.xerces.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:325)
        at org.apache.xerces.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:282)
        at org.apache.xerces.impl.xs.XMLSchemaValidator$XSIErrorReporter.reportError(XMLSchemaValidator.java:525)
        at org.apache.xerces.impl.xs.XMLSchemaValidator.reportSchemaError(XMLSchemaValidator.java:4101)
        at org.apache.xerces.impl.xs.XMLSchemaValidator.handleEndElement(XMLSchemaValidator.java:2812)
        at org.apache.xerces.impl.xs.XMLSchemaValidator.endElement(XMLSchemaValidator.java:968)
        at org.apache.xerces.jaxp.validation.ValidatorHandlerImpl.endElement(ValidatorHandlerImpl.java:608)
        at net.sf.saxon.event.ContentHandlerProxy.endElement(ContentHandlerProxy.java:414)
```

--- a/RELEASE.md
+++ b/RELEASE.md
@@ -1,4 +1,7 @@
 # Releases #
+## In Progress Work ##
+1. Fixed a bug where the removeDups opt sometimes created checkers with missing states.
+
 ## Release 2.5.0 (2017-09-25) ##
 1. Fixed a bug where method labels were sometimes lost when the rax:roles was used.
 1. Clean up : better handling of error states in optimization stages.

--- a/core/src/main/resources/xsl/opt/removeDups.xsl
+++ b/core/src/main/resources/xsl/opt/removeDups.xsl
@@ -84,9 +84,13 @@
             <xsl:otherwise>
                 <xsl:call-template name="replaceAllDups">
                     <xsl:with-param name="checker">
-                        <xsl:call-template name="replaceDups">
-                            <xsl:with-param name="checker" select="util:pruneSteps($checker)"/>
-                            <xsl:with-param name="dups" select="$dups"/>
+                        <xsl:call-template name="util:pruneSteps">
+                            <xsl:with-param name="checker">
+                                <xsl:call-template name="replaceDups">
+                                    <xsl:with-param name="checker" select="$checker"/>
+                                    <xsl:with-param name="dups" select="$dups"/>
+                                </xsl:call-template>
+                            </xsl:with-param>
                         </xsl:call-template>
                     </xsl:with-param>
                 </xsl:call-template>
@@ -112,9 +116,13 @@
             <xsl:otherwise>
                 <xsl:call-template name="replaceEpsilons">
                     <xsl:with-param name="checker">
-                        <xsl:call-template name="replaceDups">
-                            <xsl:with-param name="checker" select="util:pruneSteps($checker)"/>
-                            <xsl:with-param name="dups" select="$dups"/>
+                        <xsl:call-template name="util:pruneSteps">
+                            <xsl:with-param name="checker">
+                                <xsl:call-template name="replaceDups">
+                                    <xsl:with-param name="checker" select="$checker"/>
+                                    <xsl:with-param name="dups" select="$dups"/>
+                                </xsl:call-template>
+                            </xsl:with-param>
                         </xsl:call-template>
                     </xsl:with-param>
                 </xsl:call-template>

--- a/core/src/main/resources/xsl/util/join.xsl
+++ b/core/src/main/resources/xsl/util/join.xsl
@@ -100,10 +100,14 @@
             <xsl:otherwise>
                 <xsl:call-template name="joinDups">
                     <xsl:with-param name="checker">
-                        <xsl:call-template name="doJoin">
-                            <xsl:with-param name="checker" select="$checker"/>
-                            <xsl:with-param name="joins" select="$joins"/>
-                            <xsl:with-param name="stepsToJoin" select="check:stepsToJoin(map:keys($joins), $joins, map{})"/>
+                        <xsl:call-template name="util:pruneSteps">
+                            <xsl:with-param name="checker">
+                                <xsl:call-template name="doJoin">
+                                    <xsl:with-param name="checker" select="$checker"/>
+                                    <xsl:with-param name="joins" select="$joins"/>
+                                    <xsl:with-param name="stepsToJoin" select="check:stepsToJoin(map:keys($joins), $joins, map{})"/>
+                                </xsl:call-template>
+                            </xsl:with-param>
                         </xsl:call-template>
                     </xsl:with-param>
                 </xsl:call-template>

--- a/core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerOptDupsSpec.scala
+++ b/core/src/test/scala/com/rackspace/com/papi/components/checker/wadl/WADLCheckerOptDupsSpec.scala
@@ -1,0 +1,132 @@
+/***
+ *   Copyright 2017 Rackspace US, Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.rackspace.com.papi.components.checker.wadl
+
+import com.rackspace.com.papi.components.checker.TestConfig
+import org.junit.runner.RunWith
+import org.scalatest.junit.JUnitRunner
+
+import scala.xml._
+
+@RunWith(classOf[JUnitRunner])
+class WADLCheckerOptDupsSpec extends BaseCheckerSpec {
+
+  //
+  //  Register some common prefixes, you'll need the for XPath
+  //  assertions.
+  //
+  register ("chk","http://www.rackspace.com/repose/wadl/checker")
+  register ("xsd", "http://www.w3.org/2001/XMLSchema")
+
+  feature ("The WADLCheckerBuilder can correctly transforma a WADL into checker format") {
+
+    info ("As a developer")
+    info ("I want to be able to transform a WADL which utilizes rax roles and the remove dups option into ")
+    info ("a description of a machine that can handle the correct header validations in checker format")
+    info ("so that an API validator can process the checker format to validate the roles")
+
+    //
+    //  Note that we are testing a very specific scenario that forces
+    //  a failure if the remove dups operation is implemented incorrectly.
+    //
+    //  More comperhensive tests of RaxRoles and Remove Dups
+    //  operations are in other Specs.
+    //
+
+    val raxRolesWADL = <application xmlns:rax="http://docs.rackspace.com/api"
+         xmlns:xsd="http://www.w3.org/2001/XMLSchema"
+         xmlns="http://wadl.dev.java.net/2009/02"
+    >
+  <resources base="http://localhost">
+    <resource id="_0be44e11-68e6-44d4-acae-ec6a366282ba" path="/path/to/this">
+      <method name="GET"    id="_0be44e11-68e6-44d4-acae-ec6a366282ba-GET"    rax:roles="role1 role2 role3 role4"/>
+    </resource>
+    <resource id="_0e126d22-d8d2-4ece-8667-b19875ae1e12" path="/path/to/this">
+      <method name="PUT"    id="_0e126d22-d8d2-4ece-8667-b19875ae1e12-PUT"    rax:roles="role1 role2 role3"/>
+    </resource>
+    <resource id="_b3d2229d-cc90-4d53-9f05-c1c60117af07" path="/path/to/this">
+      <method name="POST"   id="_b3d2229d-cc90-4d53-9f05-c1c60117af07-POST"   rax:roles="role1 role2"/>
+    </resource>
+    <resource id="_4724d9ab-d560-4668-bc18-3690cc7faa76" path="/path/to/this">
+      <method name="DELETE" id="_4724d9ab-d560-4668-bc18-3690cc7faa76-DELETE" rax:roles="role1"/>
+    </resource>
+    <resource id="_bed37f49-c275-474b-8681-a3a48918ddb9" path="/path/to/that">
+      <method name="GET"    id="_bed37f49-c275-474b-8681-a3a48918ddb9-GET"    />
+      <method name="PUT"    id="_bed37f49-c275-474b-8681-a3a48918ddb9-PUT"    />
+    </resource>
+    <resource id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284" path="/path/to/that">
+      <method name="GET"    id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284-GET"    rax:roles="role1"/>
+      <method name="PUT"    id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284-PUT"    rax:roles="role1"/>
+      <method name="POST"   id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284-POST"   rax:roles="role1"/>
+      <method name="DELETE" id="_6fdeaf12-ca1d-443e-bd77-25c5cb8b0284-DELETE" rax:roles="role1"/>
+    </resource>
+  </resources>
+    </application>
+
+    val config = {
+      val tc = TestConfig()
+      tc.removeDups = true
+      tc.enableRaxRolesExtension = true
+      tc
+    }
+
+    scenario ("The WADL contains rax:roles with remove dups optimization enabled and a flow which allows a method call with and without roles") {
+      Given("A WADL contains rax:roles with remove dups optimization enabled and a flow which allows a method call with and without roles")
+      When ("The wadl is translated")
+      val checker = builder.build(raxRolesWADL, config)
+      Then("The following assertions should hold...")
+      assert(checker, "count(chk:checker/chk:step[@type='METHOD' and @match='GET'])    = 3")
+      assert(checker, "count(chk:checker/chk:step[@type='METHOD' and @match='POST'])   = 2")
+      assert(checker, "count(chk:checker/chk:step[@type='METHOD' and @match='PUT'])    = 3")
+      assert(checker, "count(chk:checker/chk:step[@type='METHOD' and @match='DELETE']) = 1")
+      assert(checker, "count(chk:checker/chk:step[@type='HEADER_ANY' and @match='role1']) = 1")
+      assert(checker, "count(chk:checker/chk:step[@type='HEADER_ANY' and @match='role1|role2']) = 1")
+      assert(checker, "count(chk:checker/chk:step[@type='HEADER_ANY' and @match='role1|role2|role3']) = 1")
+      assert(checker, "count(chk:checker/chk:step[@type='HEADER_ANY' and @match='role1|role2|role3|role4']) = 1")
+      assert(checker, "count(chk:checker/chk:step[@type='CAPTURE_HEADER' and @path='for $h in req:headers(''X-ROLES'', true()) return if ($h = (''role1'')) then $h else ()']) = 1")
+      assert(checker, "count(chk:checker/chk:step[@type='CAPTURE_HEADER' and @path='for $h in req:headers(''X-ROLES'', true()) return if ($h = (''role1'', ''role2'')) then $h else ()']) = 1")
+      assert(checker, "count(chk:checker/chk:step[@type='CAPTURE_HEADER' and @path='for $h in req:headers(''X-ROLES'', true()) return if ($h = (''role1'', ''role2'', ''role3'')) then $h else ()']) = 1")
+      assert(checker,
+             "count(chk:checker/chk:step[@type='CAPTURE_HEADER' and @path='for $h in req:headers(''X-ROLES'', true()) return if ($h = (''role1'', ''role2'', ''role3'', ''role4'')) then $h else ()']) = 1")
+
+      List("GET","PUT").foreach (m => {
+        assert(checker, Start, URL("path"), URL("to"), URL("that"), Method(m), Accept)
+      })
+
+      List("GET","PUT","POST","DELETE").foreach ( m => {
+        assert(checker, Start, URL("path"), URL("to"), URL("that"), Method(m), HeaderAny("X-ROLES","role1"),
+               RaxCaptureHeader("X-RELEVANT-ROLES", "for $h in req:headers('X-ROLES', true()) return if ($h = ('role1')) then $h else ()"),
+               Accept)
+      })
+
+      assert(checker, Start, URL("path"), URL("to"), URL("this"), Method("DELETE"), HeaderAny("X-ROLES","role1"),
+             RaxCaptureHeader("X-RELEVANT-ROLES", "for $h in req:headers('X-ROLES', true()) return if ($h = ('role1')) then $h else ()"),
+             Accept)
+
+      assert(checker, Start, URL("path"), URL("to"), URL("this"), Method("POST"), HeaderAny("X-ROLES","role1|role2"),
+             RaxCaptureHeader("X-RELEVANT-ROLES", "for $h in req:headers('X-ROLES', true()) return if ($h = ('role1', 'role2')) then $h else ()"),
+             Accept)
+
+      assert(checker, Start, URL("path"), URL("to"), URL("this"), Method("PUT"), HeaderAny("X-ROLES","role1|role2|role3"),
+             RaxCaptureHeader("X-RELEVANT-ROLES", "for $h in req:headers('X-ROLES', true()) return if ($h = ('role1', 'role2', 'role3')) then $h else ()"),
+             Accept)
+
+      assert(checker, Start, URL("path"), URL("to"), URL("this"), Method("GET"), HeaderAny("X-ROLES","role1|role2|role3|role4"),
+             RaxCaptureHeader("X-RELEVANT-ROLES", "for $h in req:headers('X-ROLES', true()) return if ($h = ('role1', 'role2', 'role3', 'role4')) then $h else ()"),
+             Accept)
+    }
+  }
+}

